### PR TITLE
Slim pallet aleph

### PR DESCRIPTION
**Pull request by mike1729**
Remove storage from pallet aleph that was used for tests and caused problems on production.

### CODE REVIEW

File: .github/workflows/build-node-and-runtime.yml [hunk 1]
```diff
@@ -44,15 +44,37 @@ jobs:
...
+      - name: Upload test runtime
+        uses: actions/upload-artifact@v2
+        with:
+          name: ${{ inputs.artifact-prefix }}aleph-test-runtime
+          path: target/release/wbuild/aleph-runtime/aleph_runtime.compact.wasm
+          if-no-files-found: error
+          retention-days: 7
+
+      - name: Stop cache
+        uses: ./.github/actions/post-cache
```
Comment (pmikolajczyk41) [inline .github/workflows/build-node-and-runtime.yml:80]: you want to remove previous step in line 41 which already stops the cache

File: .github/workflows/build-node-and-runtime.yml [hunk 2]
```diff
@@ -44,15 +44,37 @@ jobs:
       - name: Upload binary
```
Comment (pmikolajczyk41) [inline .github/workflows/build-node-and-runtime.yml:44]: just for consistency: `Upload release binary`

File: .github/workflows/e2e-tests-main-devnet.yml [hunk 3]
```diff
@@ -19,34 +19,33 @@ concurrency:
 jobs:
   build-new-node:
     name: Build node and runtime artifacts (PR version)
-    uses: Cardinal-Cryptography/aleph-node/.github/workflows/build-node-and-runtime.yml@main
-
+    uses: ./.github/workflows/build-node-and-runtime.yml
```
Comment (pmikolajczyk41) [inline .github/workflows/e2e-tests-main-devnet.yml:22]: +1

I'm quite sure that at the time of introducing reusable workflows to our repo, the only way of pointing workflow was through full git path :(

File: bin/node/Cargo.toml [hunk 4]
```diff
@@ -68,3 +68,7 @@ substrate-build-script-utils = { git = "https://github.com/paritytech/substrate.
 
 [features]
 default = []
+short_session = [
+"aleph-runtime/short_session",
+"aleph-primitives/short_session"
```
Comment (pmikolajczyk41) [inline bin/node/Cargo.toml:73]: add here some indent please

File: bin/runtime/src/lib.rs [hunk 5]
```diff
@@ -104,19 +107,21 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
     spec_name: create_runtime_str!("aleph-node"),
     impl_name: create_runtime_str!("aleph-node"),
     authoring_version: 1,
-    spec_version: 8,
+    spec_version: 9,
```
Comment (pmikolajczyk41) [inline bin/runtime/src/lib.rs:110]: most probably you will be merging after #307 so remember to bump this to `10`

File: docker/Dockerfile [hunk 6]
```diff
@@ -1,4 +1,4 @@
-FROM ubuntu:focal-20210827
+FROM ubuntu:latest
```
Comment (pmikolajczyk41) [inline docker/Dockerfile:1]: uhm, using `latest` sometimes may lead to some unpleasant surprises; if you do not have strong incentive, I would either leave the previous 'sealed' one or choose newer, but still, 'sealed'

File: pallets/aleph/Cargo.toml [hunk 7]
```diff
@@ -2,21 +2,21 @@
...
 serde = "1.0"
-primitives = { path = "../primitives", default-features = false}
-pallet-session = { default-features = false, git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.13" }
 sp-io = { default-features = false, git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.13" }
 scale-info = { version = "1.0", default-features = false, features = ["derive"] }
+primitives = { path = "../../primitives", default-features = false}
+pallet-balances = { default-features = false, git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.13" }
+pallet-session = { default-features = false, git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.13" }
 
 [dev-dependencies]
```
Comment (pmikolajczyk41) [inline pallets/aleph/Cargo.toml:19]: does this section with substrate dependencies only makes sense here?

File: pallets/aleph/src/mock.rs [hunk 8]
```diff
@@ -33,7 +31,7 @@ construct_runtime!(
         System: frame_system::{Pallet, Call, Config, Storage, Event<T>},
         Balances: pallet_balances::{Pallet, Call, Storage, Config<T>, Event<T>},
         Session: pallet_session::{Pallet, Call, Storage, Event, Config<T>},
-        Aleph: pallet_aleph::{Pallet, Call, Config<T>, Storage, Event<T>},
+        Aleph: pallet_aleph::{Pallet,  Storage},
```
Comment (pmikolajczyk41) [inline pallets/aleph/src/mock.rs:34]: one space should be enough :)

**Review by pmikolajczyk41 — CHANGES_REQUESTED** [review]
There are a lot of changes that (in my opinion) should be placed in a separate PR:
 - I would isolate all the rust-related changes (version, edition, dependencies)
 - I would isolate 'short-session' feature
 - what remains should form a more consistent PR

A general nit: until now we usually tried to keep the blank line at the end of every file (kinda unnecessary rule, but anyway followed). We should either drop it and announce that or (preferably) add the blank lines to all the files in this PR that are missing it (there are a couple of them).

(I'm still reviewing the core changes)

File: pallets/aleph/src/migrations/v1_to_v2.rs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+            log::info!(target: "pallet_aleph", "Storage item SessionPeriod does not exist!");
+        }
+        SessionPeriod::kill();
+
+        // store new version
+        StorageVersion::new(2).put::<P>();
+
+        T::DbWeight::get().reads(1) + T::DbWeight::get().writes(1)
+    } else {
+        log::warn!(
```
Comment (pmikolajczyk41) [inline pallets/aleph/src/migrations/v1_to_v2.rs:40]: why `warn` here?

File: pallets/aleph/src/mock.rs [hunk 10]
```diff
@@ -173,18 +174,26 @@ pub fn new_test_ext(authorities: &[(u64, u64)]) -> sp_io::TestExternalities {
...
-pub(crate) fn run_session(n: u64) {
-    while System::block_number() < n {
+pub(crate) fn run_session(n: u32) {
+    for i in Session::current_index()..n {
         Session::on_finalize(System::block_number());
         Aleph::on_finalize(System::block_number());
         System::on_finalize(System::block_number());
 
+        let parent_hash = if System::block_number() > 1 {
+            System::finalize().hash()
```
Comment (pmikolajczyk41) [inline pallets/aleph/src/mock.rs:184]: what are the reasons behind introducing `System::finalize().hash()` and `set_block_number` (below)?

File: pallets/elections/src/lib.rs [hunk 11]
```diff
@@ -20,7 +20,7 @@ pub mod pallet {
     use frame_election_provider_support::{
         ElectionDataProvider, ElectionProvider, Support, Supports,
     };
-    use frame_support::pallet_prelude::*;
+    use frame_support::{pallet_prelude::*, traits::Get};
```
Comment (pmikolajczyk41) [inline pallets/elections/src/lib.rs:23]: do you need this import?

File: pallets/aleph/src/tests.rs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+        Aleph::update_authorities(to_authorities(&[2, 3, 4]).as_slice());
+
+        assert_eq!(Aleph::authorities(), to_authorities(&[2, 3, 4]));
+
+        run_session(2);
+
+        Aleph::update_authorities(to_authorities(&[1, 2, 3]).as_slice());
+
+        assert_eq!(Aleph::authorities(), to_authorities(&[1, 2, 3]),);
```
Comment (pmikolajczyk41) [inline pallets/aleph/src/tests.rs:49]: remove trailing comma

File: pallets/aleph/src/tests.rs [hunk 13]
```diff
@@ -0,0 +1,64 @@
+#![cfg(test)]
```
Comment (pmikolajczyk41) [inline pallets/aleph/src/tests.rs:1]: why do we drop migration tests?

File: pallets/aleph/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+        fn on_genesis_session<'a, I: 'a>(validators: I)
+        where
+            I: Iterator<Item = (&'a T::AccountId, T::AuthorityId)>,
+            T::AccountId: 'a,
+        {
+            let (_, authorities): (Vec<_>, Vec<_>) = validators.unzip();
+            Self::initialize_authorities(authorities.as_slice());
+        }
+
+        fn on_new_session<'a, I: 'a>(_changed: bool, validators: I, _queued_validators: I)
```
Comment (pmikolajczyk41) [inline pallets/aleph/src/lib.rs:89]: why don't we take advantage of `changed` flag?

File: pallets/aleph/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+        fn on_runtime_upgrade() -> frame_support::weights::Weight {
+            migrations::v0_to_v1::migrate::<T, Self>() + migrations::v1_to_v2::migrate::<T, Self>()
+        }
+    }
+
+    #[pallet::storage]
+    #[pallet::getter(fn authorities)]
+    pub(super) type Authorities<T: Config> = StorageValue<_, Vec<T::AuthorityId>, ValueQuery>;
+
+    impl<T: Config> Pallet<T> {
```
Comment (pmikolajczyk41) [inline pallets/aleph/src/lib.rs:57]: do I understand correctly, that the newest version of this pallet is only to serve `authorities` set internally to the AlephBFT (which itself is computed elsewhere)? if so, then it seems as a massive overhead for something that seems achievable without dedicated pallet

File: pallets/aleph/src/migrations/v1_to_v2.rs [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+        if !SessionPeriod::exists() {
+            log::info!(target: "pallet_aleph", "Storage item SessionPeriod does not exist!");
+        }
+        SessionPeriod::kill();
+
+        // store new version
+        StorageVersion::new(2).put::<P>();
+
+        T::DbWeight::get().reads(1) + T::DbWeight::get().writes(1)
```
Comment (pmikolajczyk41) [inline pallets/aleph/src/migrations/v1_to_v2.rs:38]: don't we use DB when calling `exists` and `kill`?

**Review by pmikolajczyk41 — CHANGES_REQUESTED** [review]

File: pallets/aleph/src/lib.rs (hunk 15 above)
Comment (mike1729) [reply pallets/aleph/src/lib.rs:57]: Talked through :)

File: .github/workflows/build-node-and-runtime.yml (hunk 1 above)
Comment (mike1729) [reply .github/workflows/build-node-and-runtime.yml:80]: Done.

File: .github/workflows/build-node-and-runtime.yml (hunk 2 above)
Comment (mike1729) [reply .github/workflows/build-node-and-runtime.yml:44]: Changed.

File: bin/node/Cargo.toml (hunk 4 above)
Comment (mike1729) [reply bin/node/Cargo.toml:73]: Done.

File: bin/runtime/src/lib.rs (hunk 5 above)
Comment (mike1729) [reply bin/runtime/src/lib.rs:110]: bumped.

File: docker/Dockerfile (hunk 6 above)
Comment (mike1729) [reply docker/Dockerfile:1]: Changed.

File: pallets/aleph/Cargo.toml (hunk 7 above)
Comment (mike1729) [reply pallets/aleph/Cargo.toml:19]: Unfortunately, without it, cargo doesn't compile with `-p`.

File: pallets/aleph/src/mock.rs (hunk 8 above)
Comment (mike1729) [reply pallets/aleph/src/mock.rs:34]: Removed.

File: pallets/aleph/src/migrations/v1_to_v2.rs (hunk 9 above)
Comment (mike1729) [reply pallets/aleph/src/migrations/v1_to_v2.rs:40]: We don't want to see it by default on info level.

File: pallets/aleph/src/mock.rs (hunk 10 above)
Comment (mike1729) [reply pallets/aleph/src/mock.rs:184]: We don't need to run all blocks from scratch on every call of this method.

File: pallets/elections/src/lib.rs (hunk 11 above)
Comment (mike1729) [reply pallets/elections/src/lib.rs:23]: Get? Yes, it's needed for `SessionPerdiod`.

File: pallets/aleph/src/tests.rs (hunk 12 above)
Comment (mike1729) [reply pallets/aleph/src/tests.rs:49]: Done.

File: pallets/aleph/src/tests.rs (hunk 13 above)
Comment (mike1729) [reply pallets/aleph/src/tests.rs:1]: It would check nothing as `StorageVersion!=default`.

File: pallets/aleph/src/lib.rs (hunk 14 above)
Comment (mike1729) [reply pallets/aleph/src/lib.rs:89]: Good point, changed.

File: pallets/aleph/src/migrations/v1_to_v2.rs (hunk 16 above)
Comment (mike1729) [reply pallets/aleph/src/migrations/v1_to_v2.rs:38]: We do, changed.

File: pallets/aleph/src/lib.rs (hunk 15 above)
Comment (mike1729) [reply pallets/aleph/src/lib.rs:57]: This is the way.

File: pallets/aleph/src/mock.rs [hunk 17]
```diff
@@ -173,18 +174,26 @@ pub fn new_test_ext(authorities: &[(u64, u64)]) -> sp_io::TestExternalities {
     t.into()
 }
 
-pub(crate) fn run_session(n: u64) {
-    while System::block_number() < n {
+pub(crate) fn run_session(n: u32) {
+    for i in Session::current_index()..n {
         Session::on_finalize(System::block_number());
         Aleph::on_finalize(System::block_number());
         System::on_finalize(System::block_number());
```
Comment (pmikolajczyk41) [inline pallets/aleph/src/mock.rs:181]: not sure how this is relevant, but as far as I understand, the hook `on_finalize` is called on pallets in the reversed order than the one defined in `bin/runtime/src/lib.rs`;
however, we ordered Aleph after Session, so in order to mock the real behavior we should keep the true succession

**Review by Yithis — APPROVED** [review]

File: pallets/aleph/src/lib.rs [hunk 18]
```diff
@@ -46,7 +47,27 @@ pub mod pallet {
...
     impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
         fn on_runtime_upgrade() -> frame_support::weights::Weight {
-            migrations::v0_to_v1::migrate::<T, Self>() + migrations::v1_to_v2::migrate::<T, Self>()
+            let on_chain = <Pallet<T> as GetStorageVersion>::on_chain_storage_version();
+            let current = 2;
+            let prev = 1;
+            let genesis = 0;
+            T::DbWeight::get().reads(1)
+                + match on_chain {
+                    _ if on_chain == current => 0,
```
Comment (fbielejec) [inline pallets/aleph/src/lib.rs:56]: perhaps just use explicit versions, w-out storing them in a variable?

File: pallets/aleph/src/lib.rs [hunk 19]
```diff
@@ -46,7 +47,27 @@ pub mod pallet {
...
+                    _ if on_chain == prev => migrations::v0_to_v1::migrate::<T, Self>(),
```
Comment (fbielejec) [inline pallets/aleph/src/lib.rs:57]: `v1_to_v2`

File: pallets/aleph/src/mock.rs (hunk 17 above)
Comment (mike1729) [reply pallets/aleph/src/mock.rs:180]: Good catch, changed.

File: pallets/aleph/src/lib.rs (hunk 18 above)
Comment (mike1729) [reply pallets/aleph/src/lib.rs:56]: Changed.

File: pallets/aleph/src/lib.rs (hunk 19 above)
Comment (mike1729) [reply pallets/aleph/src/lib.rs]: Fixed.

File: pallets/aleph/src/mock.rs (hunk 17 above)
Comment (pmikolajczyk41) [reply pallets/aleph/src/mock.rs:180]: almost :) the `on_initialize` block below is out-of-order :stuck_out_tongue:

**Review by pmikolajczyk41 — APPROVED** [review]

**Pull request merged**